ControlWork/Queue.cs

```diff
@@ -0,0 +1,67 @@
+﻿namespace ControlWork;
+
+public class ThreadSafetyQueue<T, P>
+    where P : IComparable
+{
+    private readonly List<QueueElement<T, P>> _elements = new();
+    private readonly object _locker = new();
+    private int _size = 0;
+
+    public void Enqueue(T element, P priority)
+    {
+        lock (_locker)
+        {
+            _elements.Add(new QueueElement<T, P>(element, priority));
+        }
+
+        Interlocked.Increment(ref _size);
```

Comment:
```suggestion
        lock (_locker)
        {
            _elements.Add(new QueueElement<T, P>(element, priority));
            ++_size;
             Monitor.PulseAll(_locker);
        }
```
Мы и так залочиваемся в этой функции. Изменения этой переменной можно производить под блокировкой одного объекта. Тогда гарантировать атомарность операции не нужно.

И нужно разбудить другие потоки, которые могут ожидать добавления элемента.

Reply:
-0.5